lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java

```diff
@@ -18,198 +18,82 @@
-  private final JobJarMaker jobJarMaker;
   private final CommandBuilder launchCommandBuilder;
 
   @Inject
   AzureBatchResourceManager(
-      final LocalAddressProvider localAddressProvider,
-      final REEFFileNames fileNames,
-      final REEFEventHandlers reefEventHandlers,
       final ConfigurationSerializer configurationSerializer,
-      final AzureStorageUtil azureStorageUtil,
-      final JobJarMaker jobJarMaker,
       final CommandBuilder launchCommandBuilder,
-      @Parameter(AzureBatchAccountUri.class) final String azureBatchAccountUri,
-      @Parameter(AzureBatchAccountName.class) final String azureBatchAccountName,
-      @Parameter(AzureBatchAccountKey.class) final String azureBatchAccountKey,
+      final AzureBatchEvaluatorShimManager evaluatorShimManager,
       @Parameter(JVMHeapSlack.class) final double jvmHeapSlack) {
-    this.localAddress = localAddressProvider.getLocalAddress();
-    this.fileNames = fileNames;
-    this.reefEventHandlers = reefEventHandlers;
     this.configurationSerializer = configurationSerializer;
-    this.azureStorageUtil = azureStorageUtil;
-    this.azureBatchAccountKey = azureBatchAccountKey;
-    this.azureBatchAccountName = azureBatchAccountName;
-    this.azureBatchAccountUri = azureBatchAccountUri;
-    this.jobId = System.getenv(AZ_BATCH_JOB_ID_ENV);
-    this.taskId = "EvaluatorTask-"
-        + this.azureBatchAccountName + "-"
-        + (new Date()).toString()
-        .replace(' ', '-')
-        .replace(':', '-')
-        .replace('.', '-');
+    this.evaluatorShimManager = evaluatorShimManager;
     this.jvmHeapFactor = 1.0 - jvmHeapSlack;
-    this.jobJarMaker = jobJarMaker;
     this.launchCommandBuilder = launchCommandBuilder;
+    this.containerRequests = new ConcurrentHashMap<>();
   }
 
   public void onResourceRequested(final ResourceRequestEvent resourceRequestEvent) {
-    final String id = UUID.randomUUID().toString();
-    final int memorySize = resourceRequestEvent.getMemorySize().get();
-
-    // TODO: Investigate nodeDescriptorHandler usage and remove below dummy node descriptor.
-    this.reefEventHandlers.onNodeDescriptor(NodeDescriptorEventImpl.newBuilder()
-        .setIdentifier(id)
-        .setHostName(this.localAddress)
-        .setPort(1234)
-        .setMemorySize(memorySize)
-        .build());
-
-    this.reefEventHandlers.onResourceAllocation(ResourceEventImpl.newAllocationBuilder()
-        .setIdentifier(id)
-        .setNodeId(id)
-        .setResourceMemory(memorySize)
-        .setVirtualCores(1)
-        .setRuntimeName(RuntimeIdentifier.RUNTIME_NAME)
-        .build());
-
-    this.containers.put(id, resourceRequestEvent);
+    LOG.log(Level.FINEST, "Got ResourceRequestEvent for Id: {0} in AzureBatchResourceManager");
+    for (int r = 0; r < resourceRequestEvent.getResourceCount(); r++) {
+      final String containerId = generateContainerId();
+      this.containerRequests.put(containerId, resourceRequestEvent);
```

Comment:
So I'm not very experienced with Java concurrency, but from my understanding, since containerRequests is a ConcurrentHashMap, the synchronization is handled for us for get/put/remove, etc operations. If we need to do compound operations (i.e. get the value, if the value exists, do A, else do B, finally remove it from the map), then you do need to synchronize.
Since in your original code you only use put, get and remove; I think using ConcurrectHashMap should be good enough.

Let me know if your understanding is different, or if you find any other information regarding this subject.

Reply:
You are right. I just checked there is no need to add synchronized() in a single operation in ConcurrectHashMap.